### ISSUE TITLE
Pull request list tracking

### ISSUE THREAD

**Issue by joshaber** [desktop/desktop#2642]
* Probably a tab in the Branches foldout
* Group my pull requests vs. others

Dependencies:
- [x] Add Pull Request API #2799
- [x] Pull Request Persistance #2779 
- [x] PR List Component #2780 
- [x] CI Build Status #2643 ~#2781~
- [x] Dynamic label for Create/Open PR #2424 
- [x] Show pull request info in "Current Branch" toolbar button #2820
- [x] Select current pull request by default #2873
- [x] Load pull request CI status independently of loading PRs #3019
- [x] Refresh the current PR's CI status periodically #3020
- [x] Show PR Status description #3134

Designs TBD

**Comment by niik** [comment]
I wanted to share some mocks for this feature. I haven't had time to talk through the finer details with @donokuda so beware that things might change.

### PR list

As stated in the issue body, we anticipate this to be a tab under the branches dropdown. Unlike the issue body, however, there's no PR groupings here. I've been going back and forth on it for a while but right now I'm in the camp of let's try out not grouping these and see how it feels. My reasons are that without groups we match dotcom's ordering precisely, making for easy at-a-glance navigation. I don't expect nor intend for this feature to be about discovering pull requests but rather quickly checking our a PR branch locally after, say, having been added as a reviewer. You'd know the title of the PR but not necessarily the branch name (on desktop/desktop these are only distantly related through humor).

![33f2a5ac-008a-48b6-98d0-fa57963c7ffc](https://user-images.githubusercontent.com/634063/30381166-ecce2fe8-989b-11e7-874b-d2947c8136aa.png)

### PR loading

While the implementation details around this are very much to be defined I've assumed that we're going to be able to load the PR list very quickly. A cold load of the list should require one API call for the vast majority of cases and then N requests to get the CI status of each (oh graphql, where art thou). We'd render the list immediately and then populate CI status incrementally as we get it.

With those assumptions in mind I didn't want a dedicated loading view with a spinner as I felt that would be too jarring when we swap in the list only moments after so I took a stab at a skeleton screen instead.

![0f9ad0ad-ed8a-46ce-8703-c5fdd87fc11f](https://user-images.githubusercontent.com/634063/30381406-a1918b28-989c-11e7-87ba-ae2b58a8d0ee.png)

While the list is being loaded these would have a subtle wave/pulse animation and once the list comes in we'd cross-fade opacity between the skeleton list and the real one and enable the filter list.

### No PRs? No Problems

Following discussions about other blankslates I wanted to try for positive messaging and onward direction rather than the standard "No PRs to see here" blankslate. I'm imagining three distinct variations of this blankslate. 

#### 1. When the branch you're on is the default branch

![b1006894-a4de-4189-958b-cfe4e5bf1110](https://user-images.githubusercontent.com/634063/30381569-0d600fdc-989d-11e7-8492-d4361d8101a3.png)

#### 2. When the branch you're on is a feature branch that is not yet a PR

![067df65f-8088-448f-a08a-7265ad5283c1](https://user-images.githubusercontent.com/634063/30381599-24cec94c-989d-11e7-9887-9b88569126f5.png)

#### 4. When the branch you're on is already a PR

Either drop the lower "tip" part of the view or re-use #1.

---

I'm still working on the toolbar button interactions here, I'll probably follow with that over in #2643 and cross link here.

**Comment by j-f1** [comment]
For the loading view, I think the skeleton PR items should fade out to the bottom. For an example, this happens on the Heroku dashboard for a project.

**Comment by donokuda** [comment]
😍 I love this!

Would it make sense to add a counter for the number of opened pull requests?

<img src="https://user-images.githubusercontent.com/1174461/30386677-51124256-985f-11e7-9b4c-55f35e586ed7.png" width="624" />

That way, someone can get a sense of how much work needs to be done before clicking into the tab. I assume the counter would be more useful for projects with a lower number of pull requests.

The filter text field also looks like it has a lower opacity. Is the intention to disable it while the list is loading?

**Comment by j-f1** [comment]
> Would it make sense to add a counter for the number of opened pull requests?

The problem is that if the list of PRs hasn’t been downloaded yet, that number can’t be provided, and folks will assume it’s 0.

**Comment by donokuda** [comment]
> The problem is that if the list of PRs hasn’t been downloaded yet, that number can’t be provided, and folks will assume it’s 0.

That's a good point. One solution could be to not display a counter if the list hasn't been downloaded yet, and then display `0` if there aren't any opened pull requests.

**Comment by j-f1** [comment]
We could also pull the list of PRs on startup, so it’s always up-to-date, then somehow pull down any PR updates.

**Comment by j-f1** [comment]
If we load all of the issues, Desktop could pass `?since=<last-fetch-date>` to load any new issues every few minutes.

**Comment by joshaber** [comment]
Yeah we should definitely eagerly load and incrementally update the list 👍 💯 We mostly already have this with our existing issues store.

**Comment by joshaber** [comment]
This looks great @niik 😍

A couple questions, apologies if you just haven't gotten to these bits yet:

1. When I click the Current Branch button, what happens? Do I see the Branches tab or PR tab? Or do we remember which one you had open last?
1. Does the Current Branch button still always show your current branch, or does it ever show the checked out PR name instead?
1. What happens on non-GitHub repositories? Hide the tab bar?

**Comment by niik** [comment]
> Would it make sense to add a counter for the number of opened pull requests?

Maybe, yeah. I've been going back and forth on it but that's mostly because I'm not sure if we'll add a third tab to this view in the future (something like "Your Branches") and I didn't want to steal horizontal space unless it felt worth it. The more I think about it though I feel like it probably is useful information, allowing you to, at a glance, see that there are PRs that need attention.

> When I click the Current Branch button, what happens? Do I see the Branches tab or PRs tab? Or do we remember which one you had open last?

I'm still working on that because I believe the answer to this will be governed entirely by the answer to your second question

> Does the Current Branch button still always show your current branch, or does it ever show the checked out PR name instead?

See my previous answer and recurse indefinitely

> What happens on non-GitHub repositories? Hide the tab bar?

Yup!

**Comment by joshaber** [comment]
>See my previous answer and recurse indefinitely

![giphy-2](https://user-images.githubusercontent.com/13760/30428931-7b0c4b0a-9923-11e7-8f57-166a1cde9251.gif)

**Comment by iAmWillShepherd** [comment]
A lot of this stuff I need for issue #2524 will likely be done in this issue

**Comment by niik** [comment]
Alright, here we go with the toolbar states.

<img src="https://user-images.githubusercontent.com/634063/30554759-e50e75e8-9ca5-11e7-82eb-12be8892ec12.png" width="261" />

So we'll change the branch icon to a PR icon when the current branch is associated with an open PR. In addition will add a subtle badge with the PR number and (if available) CI status.

The quick specs

- One half spacing (5px) margin inside the badge and between the text and the status check (this is effective margin so we might have to tweak the CSS to match but let's start with 5px)
- One spacing (10px) margin between badge and arrow (this should be taken care of already)
- Badge is right aligned with one spacing (10px) margin between the badge and the toolbar button title/description
- Standard badge border radius of 3px
- `--font-size-sm` for the badge text
- `$gray-800` for badge background
- `$orange-700`, `$green-500` and `$red-600` for the CI status icons
- full white for the badge text
- 18px badge height

When we get to the implementation we might have to either make some custom paths for the icons or tweak the CSS depending on the status check condition.

> When I click the Current Branch button, what happens? Do I see the Branches tab or PRs tab? Or do we remember which one you had open last?

I think it should remember the last tab you had opened.

> Does the Current Branch button still always show your current branch, or does it ever show the checked out PR name instead?

No, we explored a ton of different alternatives here and have come to the conclusion that branch name trumps PR title *while in the client*. That in and of itself doesn't preclude having both but the designs I've explored haven't been satisfactory so far so we're going with branch name and an indication that the branch is currently associated with an open PR.

> What happens on non-GitHub repositories? Hide the tab bar?

I'll amend this to say, yes, for now. I think this is a great spot to educate about PRs for **local only** repositories and we could have a shortcut to publishing in the tab. Let's save that for later though.

**Comment by j-f1** [comment]
That looks awesome!

What should happen when someone clicks on the PR badge? What about the status icon?

**Comment by j-f1** [comment]
Also, maybe the list item with the selected PR should be displayed at the top of the branch dropdown for quicker access.

**Comment by joshaber** [comment]
This looks great 💖 

>What should happen when someone clicks on the PR badge? What about the status icon?

I have the same question, but leading even more I wanna say that I often find myself wanting to open the PR for the current branch 🌬

**Comment by j-f1** [comment]
> I often find myself wanting to open the PR for the current branch 🌬

That would probably be covered by https://github.com/desktop/desktop/issues/2524.

**Comment by niik** [comment]
> What should happen when someone clicks on the PR badge? What about the status icon?

For now, I don't think the badge should have a function. I've been going back and forth on this because it's certainly not far-fetched to imagine it would open the current PR but I'm super vary of overloading the button as well. Buttons in buttons rarely come out well. I have been exploring UI affordances for opening the current PR both in the button and in the foldout but I'm not happy with any of them at the moment and, as @j-f1 points our we will have a way (albeit not a very discoverable one) to get there.

TL;DR: Initially nothing, but we'll keep exploring ways of getting quick access to the current PR.

**Comment by shiftkey** [comment]
Also: we already have the keyboard shortcut for "Create Pull Request" and an open issue #2524 to change the button label to "Show Pull Request" if it has one associated with it.

**Comment by joshaber** [comment]
A couple questions on how we handle forks:

1. Do we display the PRs on the fork, or on its parent? Probably the parent, right?
1. What happens when I click on a PR from a fork? Check out the pulls ref or add the remote?

**Comment by joshaber** [comment]
>4. When the branch you're on is already a PR
>
>Either drop the lower "tip" part of the view or re-use #1.

I don't understand this case... if the branch you're on is a PR, then wouldn't we show the PR list?

**Comment by iAmWillShepherd** [comment]
Closing now that all work this issue tracked is now complete.

**Comment by dababy333** [comment]
yo can you make a code to make it to were funny shooter 2 on tees pure.github.io has aim lock option pls it would really be helpful thnks
